server/jetstream_test.go

```diff
@@ -10392,6 +10392,362 @@ func TestJetStreamPullConsumerMaxAckPendingRedeliveries(t *testing.T) {
+			t.Errorf("Expected no pending requests, got: %v", info.NumWaiting)
+		}
+	})
+
+	t.Run("blocking fetch", func(t *testing.T) {
+		// Create requests that take a longer time and will exhaust
+		// the number of waiting requests so that the rest will be blocked.
+		var (
+			max         = 5
+			msgCh       = make(chan *nats.Msg, max)
+			errCh       = make(chan error, max)
+			expectedMap = make(map[string]bool)
+		)
+
+		ctx, cancelFetch := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancelFetch()
+
+		for i := 0; i < max; i++ {
+			expectedMap[fmt.Sprintf("quux:%d", i)] = false
+			go func() {
+				// Can only have at most 5 inflight fetch requests so
+				// so any fetch will timeout until these are canceled.
+				// These may receive some messages but they are not going
+				// to return until the context cancels the request forcefully later.
+				msgs, err := sub.Fetch(1000, nats.Context(ctx))
+				if err != nil {
+					errCh <- err
+				}
+				for _, msg := range msgs {
+					msgCh <- msg
+				}
+			}()
+		}
+
+		// Give some time to the fetch requests to linger.
+		time.Sleep(500 * time.Millisecond)
+		info, err := sub.ConsumerInfo()
+		if err != nil {
+			t.Fatal(err)
+		}
+		if info.NumWaiting != max {
+			t.Errorf("Expected %v pull requests, got: %v", max, info.NumWaiting)
+		}
+
+		// Send max number of messages that will be received by the first batch.
+		for i := 0; i < max; i++ {
+			js.Publish(subject, []byte(fmt.Sprintf("quux:%v", i)))
+		}
+
+		ctx, done := context.WithTimeout(context.Background(), 3*time.Second)
+		defer done()
+		time.AfterFunc(500*time.Millisecond, func() {
+			js.Publish(subject, []byte("quux:5"))
+		})
+
+		// Force unblock the first batch of requestors.
+		time.AfterFunc(1*time.Second, func() {
+			cancelFetch()
+		})
+
+		var (
+			msgs = make([]*nats.Msg, 0)
+			errs = make([]error, 0)
+		)
+
+		// Create a new pull subscriber with a different bound inbox.
+		subB, err := js.PullSubscribe(subject, "durable", nats.PullMaxWaiting(5))
+		if err != nil {
+			t.Fatal(err)
+		}
+
+	Loop:
+		for {
+			select {
+			case <-ctx.Done():
+				break Loop
+			default:
+			}
+
+			// These will timeout until all the original blocking fetch requests
+			// are canceled due to reaching max number of inflight requests.
+			m, err := subB.Fetch(1, nats.MaxWait(100*time.Millisecond))
+			if err != nil {
+				errs = append(errs, err)
+			}
+			if len(m) > 0 {
+				info, _ = subB.ConsumerInfo()
+				if info.NumWaiting != 0 {
+					t.Errorf("Expected: %v, got: %v", 0, info.NumWaiting)
+				}
+			}
+			msgs = append(msgs, m...)
+			if len(msgs) > 0 {
+				break Loop
+			}
+		}
+		// Wait for first batch of requests to be canceled.
+		<-ctx.Done()
+
+		info, _ = subB.ConsumerInfo()
+		if info.NumWaiting != 0 {
+			t.Errorf("Expected: %v, got: %v", 0, info.NumWaiting)
+		}
+		if len(errs) == 0 {
+			t.Errorf("Expected at least an error, got: %v", len(errs))
+		}
+		if len(msgs) != 1 {
```

Comment:
Here unless the older sub.Fetch requests are force expired when reaching the num waiting limit, no messages will be able to be received until any of the sub.Fetch calls expire (in this example I set the deadline to 30 seconds).

Reply:
I believe that is the correct behavior assuming the server still sees interest for the reply subject for the pending request.